components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/cache/BaseCache.java

```diff
@@ -0,0 +1,396 @@
+
+        this(cacheName, false);
+    }
+
+    public BaseCache(String cacheName, boolean isTemp) {
+
+        this.cacheName = CachingConstants.LOCAL_CACHE_PREFIX + cacheName;
+        orgMgtCacheConfig = OrganizationManagementConfigUtil.getOrgMgtCacheConfig(CACHE_MANAGER_NAME, cacheName);
+        if (orgMgtCacheConfig != null) {
+            if (orgMgtCacheConfig.isDistributed()) {
+                this.cacheName = cacheName;
+            }
+            orgMgtCacheConfig.setTemporary(isTemp);
+        }
+        if (log.isDebugEnabled()) {
+            String tenantDomain = CarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+            log.debug("Cache : " + cacheName + "  is initialized for tenant domain : " + tenantDomain);
+        }
+    }
+
+    private synchronized Cache<K, V> getBaseCache() {
+
+        Cache<K, V> cache;
+        CacheManager cacheManager = Caching.getCacheManagerFactory()
+                .getCacheManager(CACHE_MANAGER_NAME);
+
+        if (cacheBuilder == null) {
+            cacheManager.removeCache(cacheName);
+            cacheBuilder = cacheManager.<K, V>createCacheBuilder(cacheName).
+                    setExpiry(CacheConfiguration.ExpiryType.ACCESSED,
+                            new CacheConfiguration.Duration(TimeUnit.SECONDS, getCacheTimeout())).
+                    setExpiry(CacheConfiguration.ExpiryType.MODIFIED,
+                            new CacheConfiguration.Duration(TimeUnit.SECONDS, getCacheTimeout())).
+                    setStoreByValue(false);
+            cache = cacheBuilder.build();
+
+            setCapacity((CacheImpl<K, V>) cache);
+            if (log.isDebugEnabled()) {
+                String tenantDomain = CarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+                log.debug("Cache : " + cacheName + "  is built with timeout value : " +
+                        getCacheTimeout() + " and capacity : " + getCapacity() +
+                        " for tenant domain : " + tenantDomain);
+            }
+        } else {
+            cache = cacheManager.getCache(cacheName);
+            setCapacity((CacheImpl<K, V>) cache);
+        }
+        return cache;
+    }
+
+    private Cache<K, V> getBaseCacheFromCacheManager() {
+
+        CacheManager cacheManager = Caching.getCacheManagerFactory()
+                .getCacheManager(CACHE_MANAGER_NAME);
+        Cache<K, V> cache = cacheManager.getCache(cacheName);
+        setCapacity((CacheImpl<K, V>) cache);
+        return cache;
+    }
+
+    /**
+     * Add a cache entry.
+     *
+     * @param key   Key which cache entry is indexed.
+     * @param entry Actual object where cache entry is placed.
+     * @param tenantDomain The tenant domain where the cache is maintained.
+     */
+    public void addToCache(K key, V entry, String tenantDomain) {
+
+        if (!isEnabled()) {
+            return;
+        }
+
+        try {
+            startTenantFlow(tenantDomain);
+            Cache<K, V> cache;
```

Comment:
shall we take this into a private method as this block is redundant is lot of places